src/confidential.rs

```diff
@@ -324,29 +519,155 @@ impl fmt::Display for Nonce {
                     write!(f, "{:02x}", b)?;
                 }
                 Ok(())
-            },
-            Nonce::Confidential(prefix, bytes) => {
-                write!(f, "{:02x}", prefix)?;
-                for b in bytes.iter() {
-                    write!(f, "{:02x}", b)?;
+            }
+            Nonce::Confidential(pk) => write!(f, "{:02x}", pk),
+        }
+    }
+}
+
+impl Default for Nonce {
+    fn default() -> Self {
+        Nonce::Null
+    }
+}
+
+impl Encodable for Nonce {
+    fn consensus_encode<S: io::Write>(&self, mut s: S) -> Result<usize, encode::Error> {
+        match *self {
+            Nonce::Null => 0u8.consensus_encode(s),
+            Nonce::Explicit(n) => {
+                1u8.consensus_encode(&mut s)?;
+                Ok(1 + n.consensus_encode(&mut s)?)
+            }
+            Nonce::Confidential(commitment) => {
+                s.write_all(&commitment.serialize())?;
+                Ok(33)
+            }
+        }
+    }
+}
+
+impl Decodable for Nonce {
+    fn consensus_decode<D: io::BufRead>(mut d: D) -> Result<Self, encode::Error> {
+        let prefix = {
+            let buffer = d.fill_buf()?;
+
+            if buffer.is_empty() {
+                return Err(encode::Error::UnexpectedEOF);
+            }
+
+            buffer[0]
+        };
+
+        match prefix {
+            0 => {
+                // consume null value prefix
+                d.consume(1);
+                Ok(Nonce::Null)
+            }
+            1 => {
+                // ignore prefix when decoding an explicit asset
+                d.consume(1);
+                let explicit = Decodable::consensus_decode(&mut d)?;
+                Ok(Nonce::Explicit(explicit))
+            }
+            p if p == 0x02 || p == 0x03 => {
+                let bytes = <[u8; 33]>::consensus_decode(&mut d)?;
+                Ok(Nonce::Confidential(
+                    PublicKey::from_slice(&bytes).map_err(secp256k1_zkp::Error::Upstream)?,
+                ))
+            }
+            p => return Err(encode::Error::InvalidConfidentialPrefix(p)),
+        }
+    }
+}
```

Comment:
I think we might be able to consolidate these `Decodable` implementations with a macro that is parameterized by:

1. the type you want to construct: `Nonce`
2. the function you need to construct the confidential value: `PublicKey::from_slice`
3. the prefix that identifies the confidential variant: `0x02` and `0x03`

Reply:
Such a macro might be even easier to implement if all the inner types implement `Encodable` / `Decodable`.